core/src/poh_service.rs

```diff
@@ -254,24 +259,38 @@ impl PohService {
                     timing.num_hashes += hashes_per_batch;
                     let mut hash_time = Measure::start("hash");
                     let should_tick = poh_l.hash(hashes_per_batch);
+                    let ideal_time = poh_l.target_poh_time(target_ns_per_tick);
                     hash_time.stop();
                     timing.total_hash_time_ns += hash_time.as_ns();
                     if should_tick {
                         // nothing else can be done. tick required.
                         return true;
                     }
                     // check to see if a record request has been sent
-                    let get_again = record_receiver.try_recv();
-                    match get_again {
-                        Ok(record) => {
+                    if let Ok(record) = record_receiver.try_recv() {
+                        // remember the record we just received as the next record to occur
+                        *next_record = Some(record);
+                        break;
+                    }
+                    // check to see if we need to wait to catch up to ideal
+                    let wait_start = Instant::now();
+                    if ideal_time <= wait_start {
```

Comment:
is it ever possible to get into a situation where we're constantly behind and won't record? I.e. something like:

1. PoH is behind the ideal, so we don't record, and hash until the next tick, break.
2.  Now we're only slightly ahead so PoHService spin waits for `X` ms.
3. Due to some combination of scheduling/overhead/the fact that the sleep in 2., isn't exact, when we get to this Poh loop again, we are behind again, so once again we don't record,

Reply:
Also might be good to discuss with @aeyakovenko, because it seems PoH is getting less and less strict here.

With these and previous changes to throttle PoH, you  need hardware that is able to run PoH much faster than the actual output tick rate of the cluster. I think @sakridge has estimated you need at least a `5-10x` faster hashing speed in order to generate our current PoH due to these throttles, but an attacker could turn off these throttles.

Reply:
> is it ever possible to get into a situation where we're constantly behind and won't record? I.e. something like:
> 
> 1. PoH is behind the ideal, so we don't record, and hash until the next tick, break.
> 2. Now we're only slightly ahead so PoHService spin waits for `X` ms.
> 3. Due to some combination of scheduling/overhead/the fact that the sleep in `2.`, isn't exact, when we get to this Poh loop again, we are behind again, so once again we don't record.
> 
> Eventually maybe you finally are able to record, and you have a huge backlog of records

record requests always cause us to stop hashing and handle all queued up records before we hash any more. look for record_receiver.try_recv()

Reply:
> `5-10x` faster hashing speed in order to generate our current PoH due to these throttles, but an attacker could turn off these throttles.

All these will increase the efficiency so that we can get closer to the peak PoH speed, do a ton of `record`'s and still hit the wallclock time of the slot. The current one was not great in that you needed a 5-10x over speed to go the speed you wanted at load.